packages/regl-worldview/src/WorldviewContext.js

```diff
@@ -240,63 +244,104 @@ export class WorldviewContext {
+            if (x < Math.floor(width) && y < Math.floor(height) && x >= 0 && y >= 0) {
+              const pixel = new Uint8Array(4);
+
+              // read pixel value from the frame buffer
+              regl.read({
+                x,
+                y,
+                width: 1,
+                height: 1,
+                data: pixel,
+              });
+
+              currentObjectId = getIdFromColor(pixel);
+              if (currentObjectId && this.getDrawPropByHitmapId(currentObjectId).object) {
+                seenIds.push(currentObjectId);
+                seenObjects.push(this.getDrawPropByHitmapId(currentObjectId));
+              }
+            }
+          } while (currentObjectId !== 0);
+          resolve(seenIds);
         });
       });
     });
   }
 
-  callComponentHandlers = (objectId: number, ray: Ray, e: MouseEvent, mouseEventName: MouseEventEnum) => {
-    this._hitmapCalls.forEach((_, component) => {
+  callComponentHandlers = (hitmapIds: Array<HitmapId>, ray: Ray, e: MouseEvent, mouseEventName: MouseEventEnum) => {
+    this._drawCalls.forEach((drawInput, component) => {
       if (component instanceof Command) {
-        component.handleMouseEvent(objectId, e, ray, mouseEventName);
+        const hitmapIdsForCommand = intersection(this._hitmapIdManager.getHitmapIdsForCommand(component), hitmapIds);
+        const mouseEvents = hitmapIdsForCommand.map((hitmapId) => this.getDrawPropByHitmapId(hitmapId));
+        component.handleMouseEvent(mouseEvents, e, ray, mouseEventName);
       }
     });
   };
 
-  _drawInput = (isHitmap?: boolean) => {
-    const drawCallsMap = isHitmap ? this._hitmapCalls : this._drawCalls;
-    const sortedDrawCalls = Array.from(drawCallsMap.values()).sort((a, b) => (a.layerIndex || 0) - (b.layerIndex || 0));
+  _drawInput = (isHitmap?: boolean, seenObjects?: Array<MouseEventObject>) => {
+    if (isHitmap) {
+      this._hitmapIdManager.reset();
+    }
 
+    const drawCalls = Array.from(this._drawCalls.values());
+    const sortedDrawCalls = drawCalls.sort((a, b) => (a.layerIndex || 0) - (b.layerIndex || 0));
     sortedDrawCalls.forEach((drawInput: DrawInput) => {
-      const { command, drawProps, instance } = drawInput;
-      if (!drawProps) {
+      const { command, children, instance, getHitmap } = drawInput;
+      if (!children) {
         return console.debug(`${isHitmap ? "hitmap" : ""} draw skipped, props was falsy`, drawInput);
       }
       const cmd = this._compiled.get(command);
       if (!cmd) {
-        return console.warn("could not find draw command for", instance.constructor.displayName);
+        return console.warn("could not find draw command for", instance ? instance.constructor.displayName : "Unknown");
+      }
+      // draw hitmap
+      if (isHitmap && getHitmap) {
+        const assignNextIdsFn: AssignNextIdsFn = (...rest) => {
+          return this._hitmapIdManager.assignNextIds(instance, ...rest);
```

Comment:
nit: maybe just use `bind`

Reply:
Bind unfortunately caused some weird flow errors, so I went with this